modules/adtrueBidAdapter.js

```diff
@@ -606,32 +611,22 @@ export const spec = {
     return bidResponses;
   },
   getUserSyncs: function (syncOptions, responses, gdprConsent, uspConsent) {
-    let syncurl = '' + publisherId;
-
-    if (gdprConsent) {
-      syncurl += '&gdpr=' + (gdprConsent.gdprApplies ? 1 : 0);
-      syncurl += '&gdpr_consent=' + encodeURIComponent(gdprConsent.consentString || '');
-    }
-    if (uspConsent) {
-      syncurl += '&us_privacy=' + encodeURIComponent(uspConsent);
-    }
-
-    // coppa compliance
-    if (config.getConfig('coppa') === true) {
-      syncurl += '&coppa=1';
-    }
-
-    if (syncOptions.iframeEnabled) {
-      return [{
-        type: 'iframe',
-        url: USER_SYNC_URL_IFRAME + syncurl
-      }];
-    } else {
-      return [{
-        type: 'image',
-        url: USER_SYNC_URL_IMAGE + syncurl
-      }];
+    if (!responses || responses.length === 0 || (!syncOptions.iframeEnabled && !syncOptions.pixelEnabled)) {
+      return [];
     }
+    return responses.filter(rsp => rsp.body && rsp.body.ext && rsp.body.ext.cookie_sync)
+      .map(rsp => rsp.body.ext.cookie_sync)
+      .reduce((a, b) => a.concat(b), [])
+      .map(({url, type}) => ({
+        type: SYNC_TYPES[type],
+        url: url +
+          '&publisherId=' + publisherId +
+          '&zoneId=' + zoneId +
+          '&gdpr=' + (gdprConsent ? (gdprConsent.gdprApplies ? 1 : 0) : 0) +
+          '&gdpr_consent=' + encodeURIComponent((gdprConsent ? gdprConsent.consentString : '')) +
+          '&us_privacy=' + encodeURIComponent((uspConsent || '')) +
+          '&coppa=' + (config.getConfig('coppa') === true ? 1 : 0)
+      }));
```

Comment:
Also, instead of filter -> map -> reduce -> map, you can accomplish this all within a single reduce.

```javascript
return responses.reduce((accum, rsp) => {
  let cookieSync = utils.deepAccess(rsp, 'body.ext.cookie_sync');
  if (cookieSync) {
    accum.push({
        type: SYNC_TYPES[cookieSync.type],
        url: cookieSync.url +
          '&publisherId=' + publisherId +
          '&zoneId=' + zoneId +
          '&gdpr=' + (gdprConsent && gdprConsent.gdprApplies ? 1 : 0) +
          '&gdpr_consent=' + encodeURIComponent((gdprConsent ? gdprConsent.consentString : '')) +
          '&us_privacy=' + encodeURIComponent((uspConsent || ''))
      });
  }
  return accum;
}, []);
```

I believe this would return the same output.

Please add a test for getUserSyncs which has many bidresponses to validate.

Something simple like:

```js
const responses = [{
	body: {
      ext: {
        cookie_sync: {
              type: 1,
              url: 'https://hb.adtrue.com/prebid/usersync?bidder=adtrue'
            }
      }
    }
}, {}, {}, {
	body: {
      ext: {
        cookie_sync: {
              type: 2,
              url: 'https://hb.adtrue.com/prebid/usersync?bidder=adtrue'
            }
      }
    }
}];
```

Expected result should be two user syncs (not 4 since we are removing the empty responses) one being type iframe, the other type image

Reply:
> Also, instead of filter -> map -> reduce -> map, you can accomplish this all within a single reduce.
> 
> ```js
> return responses.reduce((accum, rsp) => {
>   let cookieSync = utils.deepAccess(rsp, 'body.ext.cookie_sync');
>   if (cookieSync) {
>     accum.push({
>         type: SYNC_TYPES[cookieSync.type],
>         url: cookieSync.url +
>           '&publisherId=' + publisherId +
>           '&zoneId=' + zoneId +
>           '&gdpr=' + (gdprConsent && gdprConsent.gdprApplies ? 1 : 0) +
>           '&gdpr_consent=' + encodeURIComponent((gdprConsent ? gdprConsent.consentString : '')) +
>           '&us_privacy=' + encodeURIComponent((uspConsent || ''))
>       });
>   }
>   return accum;
> }, []);
> ```
> 
> I believe this would return the same output.
> 
> Please add a test for getUserSyncs which has many bidresponses to validate.
> 
> Something simple like:
> 
> ```js
> const responses = [{
> 	body: {
>       ext: {
>         cookie_sync: {
>               type: 1,
>               url: 'https://hb.adtrue.com/prebid/usersync?bidder=adtrue'
>             }
>       }
>     }
> }, {}, {}, {
> 	body: {
>       ext: {
>         cookie_sync: {
>               type: 2,
>               url: 'https://hb.adtrue.com/prebid/usersync?bidder=adtrue'
>             }
>       }
>     }
> }];
> ```
> 
> Expected result should be two user syncs (not 4 since we are removing the empty responses) one being type iframe, the other type image

You can optimize the several loops into a single one.

Reply:
@robertrmartinez done, thanks you.